### PR TITLE
Fixed #28399 -- Fixed QuerySet.count() for union(), difference(), and intersection() queries.

### PR DESCRIPTION
**Pull request by apollo13**

### CODE REVIEW

**Comment by apollo13** [comment]
buildbot, test on oracle.

**Comment by charettes** [comment]
Since `get_aggregation()` also powers the `aggregate()` method it would be great to see if it makes much effort to get `aggregate(Avg('num'))` to work as well!

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -411,12 +411,12 @@ def get_aggregation(self, using, added_aggregate_names):
         # aren't smart enough to remove the existing annotations from the
         # query, so those would force us to use GROUP BY.
         #
-        # If the query has limit or distinct, then those operations must be
-        # done in a subquery so that we are aggregating on the limit and/or
+        # If the query has limit, distinct or uses set operations, then those operations
```
Comment (timgraham) [inline django/db/models/sql/query.py]: `# If the query has limit or distinct, or uses set operations,` (also wrap the comment at 79 chars)

File: tests/queries/test_qs_combinators.py [hunk 2]
```diff
@@ -89,6 +91,14 @@ def test_ordering(self):
         qs2 = Number.objects.filter(num__gte=2, num__lte=3)
         self.assertNumbersEqual(qs1.union(qs2).order_by('-num'), [3, 2, 1, 0])
 
+    def test_count(self):
+        qs1 = Number.objects.filter(num__lte=1)
+        qs2 = Number.objects.filter(num__gte=2, num__lte=3)
+        with CaptureQueriesContext(connection) as captured_queries:
+            self.assertEqual(qs1.union(qs2).count(), 4)
+        self.assertEqual(len(captured_queries), 1)
+        self.assertTrue(captured_queries[0]['sql'].upper().startswith("SELECT COUNT(*) FROM ("))
```
Comment (timgraham) [inline tests/queries/test_qs_combinators.py]: As that fails on Oracle and might be brittle for third-party DBs, I think it could probably be removed?

**Comment by apollo13** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
Add tests for `intersection()` and `difference()`?

**Comment by apollo13** [comment]
Done, although the usefulness of those tests is somewhat questionable since the executed SQL is basically the same.

**Comment by apollo13** [comment]
mhm, good catch. Updated per your suggestion and pushed

**Pull request merged**